neurobooth_os/iout/lsl_streamer.py

```diff
@@ -1,163 +1,291 @@
-                    del streams[kdev]
-
-    elif node_name == "presentation":
-        from neurobooth_os.iout import marker_stream
-        from neurobooth_os.iout.mouse_tracker import MouseStream
-        from neurobooth_os.iout.eyelink_tracker import EyeTracker
-
-        streams["marker"] = marker_stream()
-
-        for kdev, argsdev in kwarg_alldevs.items():
-            if "Eyelink" in kdev:
-                logger.debug(f'LSL Streamer Starting: {kdev}')
-                streams["Eyelink"] = EyeTracker(win=win, **argsdev)
-            elif "Mouse" in kdev:
-                logger.debug(f'LSL Streamer Starting: {kdev}')
-                streams["mouse"] = MouseStream(**argsdev)
-                streams["mouse"].start()
-
-            elif any([d in kdev for d in ["Mbient_LF", "Mbient_RF"]]):
-                logger.debug(f'LSL Streamer Starting: {kdev}')
-                streams[kdev] = Mbient(**argsdev)
-                if not streams[kdev].prepare():
-                    del streams[kdev]
-                else:
-                    streams[kdev].start()
+            for f in as_completed(futures):
+                f.result()  # Raise errors that occur asynchronously
+
+        self.logger.info(f'LOADED DEVICES: {list(self.streams.keys())}')
+
+    @staticmethod
+    def _get_device_kwargs(collection_id: str, conn=None) -> Dict[str, Any]:
+        """
+        Fetch the keyword arguments for each device from the database.
+
+        :param collection_id: Name of study collection in the database.
+        :param win: PsychoPy window
+        :param conn: Connection to the database
+        """
+        # Get params from all tasks
+        kwarg_devs = meta._get_device_kwargs_by_task(collection_id, conn)
+
+        # Get all device params from session
+        kwarg_alldevs = {}
+        for dc in kwarg_devs.values():
+            kwarg_alldevs.update(dc)
+
+        return kwarg_alldevs
 
-    elif node_name == "dummy_acq":
-        from neurobooth_os.mock import mock_device_streamer as mock_dev
-        from neurobooth_os.iout.iphone import IPhone
+    # TODO: The below device-specific calls should all be refactored so that devices can be generically handled
+    # TODO: E.g., devices should be able to register handlers for lifecycle phases
 
-        for kdev, argsdev in kwarg_alldevs.items():
-            if "Intel" in kdev:
-                streams[kdev] = mock_dev.MockCamera(**argsdev)
-            elif "Mbient" in kdev:
-                streams[kdev] = mock_dev.MockMbient(**argsdev)
-                streams[kdev].start()
-            elif "IPhone" in kdev:
-                success = False
-                streams[kdev] = IPhone(name="IPhoneFrameIndex", **argsdev)
-                success = streams[kdev].prepare()
-                if not success and streams.get(kdev) is not None:
-                    del streams[kdev]
+    @staticmethod
+    def is_camera(stream_name: str) -> bool:
```

Comment:
would it be better to give each device a "type" so that we don't need a list of names?

Reply:
Ideally, such checks will no longer be necessary if a proper life cycle is defined and the guarded code is placed in the correct lifecycle call.